tests/loader_test_vectors.rs

```diff
@@ -0,0 +1,231 @@
+#[test_resources("ion-tests/iontestdata/bad/**/*.10n")]
+fn bad(file_name: &str) {
+    assert_file(ALL_SKIP_LIST, file_name, || {
+        match load_file(&loader(), file_name) {
+            Ok(items) => panic!("Expected error, got: {:?}", items),
+            Err(_) => Ok(()),
+        }
+    });
+}
+
+/// Parses the elements of a given sequence as text Ion data and tests a grouping on the loaded
+/// documents.
+///
+/// For example, for the given group:
+///
+/// ```plain
+/// embedded_documents::(
+///   "a b c"
+///   "'a' 'b' 'c'"
+/// )
+/// ```
+///
+/// This will parse each string as a [`Vec`] of [`Element`] and apply the `group_assert` function
+/// for every pair of the parsed data including the identity case (a parsed document is
+/// compared against itself).
+fn load_group_embedded<L, S, F>(loader: &L, raw_group: &S, group_assert: &F) -> IonResult<()>
+where
+    L: Loader,
+    S: Sequence,
+    F: Fn(&Vec<OwnedElement>, &Vec<OwnedElement>) -> (),
+{
+    let group_res: IonResult<Vec<_>> = raw_group
+        .iter()
+        .map(|elem| loader.load_all(elem.as_str().unwrap().as_bytes()))
+        .collect();
+    let group = group_res?;
+    for this in group.iter() {
+        for that in group.iter() {
+            group_assert(this, that);
+        }
+    }
+    Ok(())
+}
+
+/// Parses a document that has top-level `list`/`sexp` values that represent a *group*.
+/// If this top-level value is annotated with `embedded_documents`, then [`load_group_embedded`]
+/// is executed for that grouping.  Otherwise, the `value_assert` function is invoked for
+/// every pair of values in a group including the identity case (a value in a group is compared
+/// against itself).
+///
+/// Here is an example with two groups that might be used for equivalence testing:
+///
+/// ```plain
+/// (name $2 'name')
+/// embedded_documents::(
+///   "a name"
+///   "a $2"
+///   "'a' 'name'"
+/// )
+/// ```
+///
+/// This would have two groups, one with direct values that will be compared and another
+/// with embedded Ion text that will be parsed and compared.
+fn load_group<L, F1, F2>(
+    loader: L,
+    file_name: &str,
+    value_assert: F1,
+    group_assert: F2,
+) -> IonResult<()>
+where
+    L: Loader,
+    F1: Fn(&OwnedElement, &OwnedElement) -> (),
+    F2: Fn(&Vec<OwnedElement>, &Vec<OwnedElement>) -> (),
+{
+    let group_lists = load_file(&loader, file_name)?;
+    for group_list in group_lists.iter() {
+        // every grouping set is a list/sexp
+        // look for the embbedded annotation to parse/test as the underlying value
+        let is_embedded = group_list
+            .annotations()
+            .find(|annotation| annotation.text() == Some("embedded_documents"))
+            .is_some();
+        match group_list.as_sequence() {
+            Some(group) => {
+                if is_embedded {
+                    load_group_embedded(&loader, group, &group_assert)?;
+                } else {
```

Comment:
This would be a bit more idiomatic as:

```rust
match (group_list.as_sequence(), is_embedded) {
  Some(group, true) => load_group_embedded(&loader, group, &group_assert)?;
  Some(group, false) => { /* ... */ }
  _ => return decoding_error( /* ... */ )
  /* ... */
}
```

Reply:
Done